### PR TITLE
Check vnet role assignment

### PR DESCRIPTION
**Pull request by gavinfish**
---

Check whether customers have granted right permission for vnet.

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/spring-cloud/azext_spring_cloud/_validators.py [hunk 1]
```diff
@@ -182,6 +182,16 @@ def validate_vnet(cmd, namespace):
             vnet_obj and vnet_obj.address_space and vnet_obj.address_space.address_prefixes \
             else '10.234.0.0/16,10.244.0.0/16,172.17.0.1/16'
 
+    vnet_permission_ready = False
+    assignments = _get_authorization_client(cmd.cli_ctx).role_assignments.list_for_scope(vnet_id)
+    for assignment in list(assignments):
+        if assignment.principal_id.lower() == "d2531223-68f9-459e-b225-5592f90d145e" and \
```
Comment (yuwzho) [inline src/spring-cloud/azext_spring_cloud/_validators.py:188]: principal id should be different for each tenant

File: src/spring-cloud/azext_spring_cloud/_validators.py [hunk 2]
```diff
@@ -182,6 +182,16 @@ def validate_vnet(cmd, namespace):
             vnet_obj and vnet_obj.address_space and vnet_obj.address_space.address_prefixes \
             else '10.234.0.0/16,10.244.0.0/16,172.17.0.1/16'
 
+    vnet_permission_ready = False
+    assignments = _get_authorization_client(cmd.cli_ctx).role_assignments.list_for_scope(vnet_id)
+    for assignment in list(assignments):
+        if assignment.principal_id.lower() == "d2531223-68f9-459e-b225-5592f90d145e" and \
+                "8e3af657-a8ff-443c-a75c-2fe8c4bcb635" in assignment.role_definition_id:
```
Comment (yuwzho) [inline src/spring-cloud/azext_spring_cloud/_validators.py:189]: How about other cloud?

Comment (gavinfish) [reply src/spring-cloud/azext_spring_cloud/_validators.py]: Any special for other cloud?

File: src/spring-cloud/azext_spring_cloud/_validators.py [hunk 3]
```diff
@@ -182,6 +182,18 @@ def validate_vnet(cmd, namespace):
             vnet_obj and vnet_obj.address_space and vnet_obj.address_space.address_prefixes \
             else '10.234.0.0/16,10.244.0.0/16,172.17.0.1/16'
 
+    graph_client = get_graph_rbac_management_client(cmd.cli_ctx)
+    target_service_principals = list(graph_client.service_principals.list(
+        filter="servicePrincipalNames/any(c:c eq '{}')".format("e8de9221-a19c-4c81-b814-fd37c6caf9d2")))
```
Comment (yuwzho) [inline src/spring-cloud/azext_spring_cloud/_validators.py:187]: Name or id? The name can be written as any string, right?

File: src/spring-cloud/azext_spring_cloud/_validators.py [hunk 4]
```diff
@@ -182,6 +182,18 @@ def validate_vnet(cmd, namespace):
...
+    target_service_principals = list(graph_client.service_principals.list(
+        filter="servicePrincipalNames/any(c:c eq '{}')".format("e8de9221-a19c-4c81-b814-fd37c6caf9d2")))
+    if not target_service_principals:
+        raise CLIError('cannot find Azure Spring Cloud Resource Provider service principal.')
+    assignments = _get_authorization_client(cmd.cli_ctx).role_assignments.list_for_scope(vnet_id)
+    vnet_permission_ready = any(assignment.principal_id.lower() == target_service_principals[0].object_id and
+                                "8e3af657-a8ff-443c-a75c-2fe8c4bcb635" in assignment.role_definition_id for assignment
+                                in assignments)
+    if not vnet_permission_ready:
+        logger.warning("Please make sure to grant Azure Spring Cloud service permission to the virtual network!")
```
Comment (yuwzho) [inline src/spring-cloud/azext_spring_cloud/_validators.py:195]: Add the doc link to the error message? You may need to find Asir to create a short link for the docs.

File: src/spring-cloud/azext_spring_cloud/_validators.py (hunk 3 above)
Comment (gavinfish) [reply src/spring-cloud/azext_spring_cloud/_validators.py]: name here is an array which also contains id, use appid to reduce confusion.

File: src/spring-cloud/azext_spring_cloud/_validators.py (hunk 4 above)
Comment (gavinfish) [reply src/spring-cloud/azext_spring_cloud/_validators.py:195]: We need a link to the official page instead of github page?

Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_validators.py:195]: My concern is that the user has no idea about the assignment, we need to guide them to an instruction doc.

File: src/spring-cloud/azext_spring_cloud/_validators.py (hunk 2 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_validators.py]: The sp only exists in the management.azure.com? Should we skip this check when the cloud set to another?

File: src/spring-cloud/azext_spring_cloud/_validators.py (hunk 4 above)
Comment (gavinfish) [reply src/spring-cloud/azext_spring_cloud/_validators.py:195]: Yes, let's wait for the official documentation.

File: src/spring-cloud/azext_spring_cloud/_validators.py (hunk 2 above)
Comment (gavinfish) [reply src/spring-cloud/azext_spring_cloud/_validators.py]: If we support vnet feature in other cloud, we should enable this check. And if not, I think we should skip all the validation for vnet.

**Pull request merged**